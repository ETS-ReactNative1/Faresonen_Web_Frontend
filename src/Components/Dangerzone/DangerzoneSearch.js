import React, { Component } from "react";
import PropTypes from "prop-types";
import { Grid, Form, List } from "semantic-ui-react";

class DangerzoneSearch extends Component {
  constructor(props) {
    super(props);
    this.state = {
      data: {
        eliteserien: [],
        obosligaen: []
      },
      loading: true,
      search: ""
    };
  }

  componentWillReceiveProps(nextProps) {
    const { eliteserien, obosligaen } = nextProps.players;
    this.setState({
      data: {
        eliteserien,
        obosligaen
      }
    });
  }

  formatPlayers = players => (
    <List.List>
      <List.Item>
        {players.map(p => (
          <List.Content key={p.name}>
            <List.Description>
              {p.name} - {p.value1} gule kort
            </List.Description>
          </List.Content>
        ))}
      </List.Item>
    </List.List>
  );

  createTeamElements = teams =>
    teams.map(team => {
      if (team.players.length === 0) {
        return "";
      }
      return (
        <List.Item key={team.team}>
          <List.Header>{team.team}</List.Header>
          <List.Description>{team.players.length} spillere</List.Description>
          {this.formatPlayers(team.players)}
        </List.Item>
      );
    });

  handleSearch = event => {
    const { value: search } = event.target;
    this.setState({ search });
  };

  searchFilter(search, leagueId) {
    if (this.state.data[leagueId] === undefined)
      throw new Error(
        `Wrong leagueId for ${leagueId}. Valid id's are ${Object.keys(
          this.state.data
        ).join(",")}`
      );
    const filteredPlayers = [];
    this.state.data[leagueId].forEach(team => {
      const filtered = team.players.filter(p =>
        p.name.toLowerCase().includes(search.toLowerCase())
      );
      filteredPlayers.push({ team: team.team, players: filtered });
    });
    return filteredPlayers;
  }

  render() {
    const eliteserienSearchFilter = this.searchFilter(
      this.state.search,
      "eliteserien"
    );
    const eliteserieElements = this.createTeamElements(eliteserienSearchFilter);

    const obosLigaenSearchFilter = this.searchFilter(
      this.state.search,
      "obosligaen"
    );
    const obosligaElements = this.createTeamElements(obosLigaenSearchFilter);

    return (
      <div>
        <Grid columns={1}>
          <Form>
            <Form.Field>
              <label>SÃ¸k etter spiller</label>
              <input
                placeholder="Fornavn/etternavn"
                onChange={this.handleSearch}
              />
            </Form.Field>
          </Form>
        </Grid>
        <Grid columns={2}>
          <Grid.Column>
            <h1>Eliteserien</h1>
            <List>{eliteserieElements}</List>
          </Grid.Column>
          <Grid.Column>
            <h1>OBOSligaen</h1>
            <List>{obosligaElements}</List>
          </Grid.Column>
        </Grid>
      </div>
    );
  }
}
DangerzoneSearch.propTypes = {
  players: PropTypes.shape({
    eliteserien: PropTypes.arrayOf(
      PropTypes.shape({
        team: PropTypes.string.isRequired,
        players: PropTypes.arrayOf(
          PropTypes.shape({
            name: PropTypes.string.isRequired,
            value1: PropTypes.number.isRequired
          })
        ).isRequired
      })
    ).isRequired
  }).isRequired
};

export default DangerzoneSearch;
