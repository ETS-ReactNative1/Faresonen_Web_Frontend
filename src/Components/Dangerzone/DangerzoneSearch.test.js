import React from "react";
import { shallow } from "enzyme";
import DangerzoneSearch from "./DangerzoneSearch";

function instantiateShallow() {
  return shallow(
    <DangerzoneSearch
      players={{
        eliteserien: [
          {
            team: "Brann",
            players: [
              {
                name: "Sindre",
                value1: 2
              },
              {
                name: "Ole",
                value1: 1
              }
            ]
          }
        ],
        obosligaen: [
          {
            team: "Mjøndalen",
            players: [
              {
                name: "Mads",
                value1: 2
              },
              {
                name: "Christian",
                value1: 4
              }
            ]
          }
        ]
      }}
    />
  );
}

describe("<DangerzoneSearch />", () => {
  it("Should return correct players when searching", () => {
    const elem = instantiateShallow();
    elem.instance().componentWillReceiveProps({
      players: {
        eliteserien: [
          {
            team: "Brann",
            players: [
              {
                name: "Sindre",
                value1: 2
              },
              {
                name: "Ole",
                value1: 1
              }
            ]
          }
        ],
        obosligaen: [
          {
            team: "Mjøndalen",
            players: [
              {
                name: "Mads",
                value1: 2
              },
              {
                name: "Christian",
                value1: 4
              }
            ]
          }
        ]
      }
    });
    let results = elem.instance().searchFilter("Sindre", "eliteserien");
    expect(results).toEqual([
      { team: "Brann", players: [{ name: "Sindre", value1: 2 }] }
    ]);

    results = elem.instance().searchFilter("sin", "eliteserien");
    expect(results).toEqual([
      { team: "Brann", players: [{ name: "Sindre", value1: 2 }] }
    ]);

    results = elem.instance().searchFilter("mad", "obosligaen");
    expect(results).toEqual([
      { team: "Mjøndalen", players: [{ name: "Mads", value1: 2 }] }
    ]);

    results = elem.instance().searchFilter("", "obosligaen");
    expect(results).toEqual([
      {
        team: "Mjøndalen",
        players: [
          {
            name: "Mads",
            value1: 2
          },
          {
            name: "Christian",
            value1: 4
          }
        ]
      }
    ]);
  });

  it("Should throw an error user input wrong leagueId", () => {
    const elem = instantiateShallow();
    function search() {
      elem.instance().searchFilter("Sindre", 1);
    }
    expect(search).toThrow();
  });
});
