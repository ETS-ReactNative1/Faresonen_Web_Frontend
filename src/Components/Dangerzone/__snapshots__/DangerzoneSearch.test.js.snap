// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DangerzoneSearch /> Should render properly 1`] = `
<div>
  <div
    className="ui one column grid"
  >
    <form
      action={undefined}
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        className="field"
      >
        <label>
          SÃ¸k etter spiller
        </label>
        <input
          onChange={[Function]}
          placeholder="Fornavn/etternavn"
        />
      </div>
    </form>
  </div>
  <div
    className="ui two column grid"
  >
    <div
      className="column"
    >
      <h1>
        Eliteserien
      </h1>
      <div
        className="ui list"
        role="list"
      >
        <div
          className="ui info message"
        >
          <div
            className="content"
          >
            <div
              className="header"
            >
              Ingen spillere i faresonen
            </div>
            <p>
              Ingen spillere i faresonen. Spillere i faresonen har partall gule kort.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="column"
    >
      <h1>
        OBOSligaen
      </h1>
      <div
        className="ui list"
        role="list"
      >
        <div
          className="ui info message"
        >
          <div
            className="content"
          >
            <div
              className="header"
            >
              Ingen spillere i faresonen
            </div>
            <p>
              Ingen spillere i faresonen. Spillere i faresonen har partall gule kort.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
